### ISSUE TITLE
Help with LDEF resources

### ISSUE THREAD

**Issue by antscode** [autc04/Retro68#41]
I've been trying to create a custom List Definition (LDEF/CODE resource) for displaying icons in a list, but my compiled app crashes when attempting to access the resource. If I use the standard list definition (i.e. a text label), everything works correctly.

I can put together a simple test app, but I thought I'd first see if you've had any experience with LDEF resources in Retro68?

The process I've been using to create the LDEF resource is as follows:

1. Create the necessary functions in a standalone app using add_application via CMAKE, then compile
2. Using Derez in MPW I extract the "Main" CODE resource and rename it to an LDEF
3. I then add the resulting .r file to my primary app and specify my LDEF resource ID when calling LNew

If I specify zero as the resource ID (i.e. use default text label) it works, but if I specify my custom resource ID it crashes.

Any thoughts?

**Comment by antscode** [comment]
I got it working, although I did so by compiling the LDEF resource using Think C on a Mac instead of Retro68. I then Derez'd the resource and brought it back into my Retro68 project.

I was also bitten by not using the pascal keyword at the start of my functions at first.

I'm happy, but is there a way to compile a standalone CODE resource using Retro68?

**Comment by autc04** [comment]
Have a look at the SystemExtension example program, which builds an 'INIT' code resource"

The important points are:
* A single CODE resource from an application is not suitable for use as an LDEF, as it has a header and expects to be part of a multi-segment application where everything is set up in a certain way.
* You can pass -Wl,—-mac-flat to the linker in order to get the contents of your LDEF resoiurce in a file (data fork).
* You can include this using Rez.

There’s one more problem which is not covered by the code example:
* Different from a normal application, an LDEF takes arguments.

Try the following:
```c
#include “Retro68Runtime.h”

extern "C" // if you are writing C++
pascal void MyListDefinition(short lMessage, Boolean lSelect, Rect *lRect, Cell lCell, short lDataOffset, short lDataLen, ListHandle lHandle)
{
	RETRO68_RELOCATE(); // <- This is important. Otherwise you’ll crash the first time you call another function or access a global.
	//…

	// Retro68FreeGlobals() 
}
```

You still have to tell the linker to use this as the entry point for your code resource, so you need to pass `-Wl,—mac-flat —Wl,-eMYLISTDEFINITION` to the linker command line.

---

Also note that the ancient Mac programmers had a trick: the "10-byte code resource stub trick".
http://www.mactech.com/articles/mactech/Vol.10/10.01/Jan94Tips/index.html

In short, make your LDEF contain the hex bytes `2F3A 0004 4E75 0000 0000`.  In your application, before you create your list, load the resource and replace the last four bytes by a pointer to a function in your program. The hex code will forward the call to your function.

**Comment by antscode** [comment]
So much great information thank you, I don't know how I missed the SystemExtension example.

I had initially tried the 10-byte code resource stub trick to avoid using a separate code resource, but it  would die as soon as I assigned the pointer:

    ldef = (LDEF_handle)NewHandle(sizeof(LDEF_record));
    if (ldef != NIL) {
        (**ldef).instruction = 0x4EF9;  /* JMP instruction */
        (**ldef).function = myListDef;
        LIST.listDefProc = (Handle)ldef; /* CRASH */
    }

But with this new information at hand I'll give it another go.

Thank you.

**Comment by antscode** [comment]
Sorry, can you tell me how i would use Rez in conjunction with the add_application function to add my LDEF resource?

I've gotten this far in my CMakeLists.txt:

    add_executable(IconListDef
        IconListDef.c
        IconListDef.r)
    
    set_target_properties(IconListDef PROPERTIES
        OUTPUT_NAME IconListDef.flt
    
        # set a linker flag that says we want a flat piece
         # of code in a data file
        LINK_FLAGS -Wl,--mac-flat)
    
    /* Can I still use add_application? Or do I need to replace it with add_custom_command? */
    add_application(MacHub
        MacHub.cpp
        size.r
        ....
    )

**Comment by autc04** [comment]
Finally figured this out.
It needed a small fix to `add_application.cmake` and another to `Elf2Mac` (`-e` wasn't working properly).

And then I got carried away writing a sample program, which is now part of the Samples. It illustrates both the 10-byte-stub trick and "real" code resources, but for WDEFs instead of LDEFs. Also, it's a standard Mac sample program involving windows and menus, so that was high time, anyway.

**Comment by antscode** [comment]
Thanks so much for this. The new sample is great and very well documented thank you - I also like the BeOS custom windows!

Unfortunately with the latest build, my own app is now crashing on launch. Chances are it's my own stupidity, so I'll try and diagnose it myself. Very keen to try out the 10-byte-stub trick with my LDEF.

Thanks again.

**Comment by antscode** [comment]
I raised a new issue for my crashes - but in regard to using the 10-byte-stub trick with an LDEF I have success!

![prefs](https://user-images.githubusercontent.com/131975/34659506-8195d3b8-f48d-11e7-8dfc-47e9de6921ac.png)

Thanks for all your help.
